ee/tools/puppet/fleetdm/lib/puppet/functions/fleetdm/preassign_profile.rb

```diff
@@ -12,28 +12,56 @@
   end
 
   def preassign_profile(profile_identifier, host_uuid, template, group = 'default', ensure_profile = 'present')
+    # initialize our response
+    preassign_profile_response = { 'error' => '', 'resource_changed' => false }
+
     client = Puppet::Util::FleetClient.instance
     env = closure_scope['server_facts']['environment']
     run_identifier = "#{closure_scope.catalog.catalog_uuid}-#{Puppet[:node_name_value]}"
-    response = client.preassign_profile(run_identifier, host_uuid, template, group, ensure_profile, closure_scope['environment'])
 
-    if response['error'].empty?
-      base64_checksum = Digest::MD5.base64digest(template)
-      host = client.get_host_by_identifier(host_uuid, env)
-      host_profiles = client.get_host_profiles(host['body']['host']['id'], env)
+    # initiate the pre-assignment process with fleet server
+    client_resp = client.preassign_profile(run_identifier, host_uuid, template, group, ensure_profile, closure_scope['environment'])
+    unless client_resp && client_resp['error'].empty?
+      Puppet.err("error pre-assigning profile #{profile_identifier} (ensure #{ensure_profile}): #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
 
-      if host_profiles['error'].empty?
-        Puppet.info("successfully pre-set profile #{profile_identifier} as #{ensure_profile}")
+    # get host by idenfifier to get the host id
+    client_resp = client.get_host_by_identifier(host_uuid, env)
+    unless client_resp && client_resp['error'].empty?
+      Puppet.err("error getting host by identifier #{host_uuid}: #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+    unless client_resp['body'] && client_resp['body']['host'] && client_resp['body']['host']['id']
+      Puppet.err("no host found for #{host_uuid} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+
+    # get host profiles currently assigned to the host
+    client_resp = client.get_host_profiles(client_resp['body']['host']['id'], env)
+    unless client_resp && client_resp['error'].empty?
+      Puppet.err("error getting host profiles for #{host_uuid}: #{client_resp&.[]('error')} \n\n #{template}")
+      preassign_profile_response['error'] = client_resp['error']
+      return preassign_profile_response
+    end
+    unless client_resp['body'] && client_resp['body']['profiles'] && !client_resp['body']['profiles'].empty?
+      Puppet.err("no profiles returned for #{host_uuid} \n\n #{template}")
```

Comment:
this returns an error on the first run on a device (because it doesn't have any profiles assigned yet)

should we mark the resource as changed and pre-assign the profile instead?

Reply:
@roperzh should this case always be marked as changed or does it depend on whether ensure present is set?

Reply:
@gillespi314 good thinking, I think only if ensure present is `true`, otherwise the device is already in the desired state, right? This also applies to marking the resource as "changed" (the return value)